web3/providers/eth_tester/provider.py

```diff
@@ -0,0 +1,379 @@
+    keccak,
+    decode_hex,
+    encode_hex,
+)
+
+from web3.providers import (
+    BaseProvider,
+)
+
+from eth_tester.exceptions import (
+    BlockNotFound,
+    FilterNotFound,
+    TransactionNotFound,
+    ValidationError,
+)
+
+from eth_tester.utils.formatting import (
+    apply_formatter_if,
+)
+
+from .middleware import (
+    ethereum_tester_middleware,
+    ethereum_tester_fixture_middleware,
+)
+
+
+def not_implemented(*args, **kwargs):
+    raise NotImplementedError("RPC method not implemented")
+
+
+@curry
+def call_eth_tester(fn_name, eth_tester, fn_args, fn_kwargs=None):
+    if fn_kwargs is None:
+        fn_kwargs = {}
+    return getattr(eth_tester, fn_name)(*fn_args, **fn_kwargs)
+
+
+def without_eth_tester(fn):
+    # workaround for: https://github.com/pytoolz/cytoolz/issues/103
+    # @functools.wraps(fn)
+    def inner(eth_tester, params):
+        return fn(params)
+    return inner
+
+
+def without_params(fn):
+    # workaround for: https://github.com/pytoolz/cytoolz/issues/103
+    # @functools.wraps(fn)
+    def inner(eth_tester, params):
+        return fn(eth_tester)
+    return inner
+
+
+@curry
+def preprocess_params(eth_tester, params, preprocessor_fn):
+    return eth_tester, preprocessor_fn(params)
+
+
+def static_return(value):
+    def inner(*args, **kwargs):
+        return value
+    return inner
+
+
+def client_version(eth_tester, params):
+    # TODO: account for the backend that is in use.
+    from eth_tester import __version__
+    return "EthereumTester/{version}/{platform}/python{v.major}.{v.minor}.{v.micro}".format(
+        version=__version__,
+        v=sys.version_info,
+        platform=sys.platform,
+    )
+
+
+@curry
+def null_if_excepts(exc_type, fn):
+    return excepts(
+        exc_type,
+        fn,
+        static_return(None),
+    )
+
+
+null_if_block_not_found = null_if_excepts(BlockNotFound)
+null_if_transaction_not_found = null_if_excepts(TransactionNotFound)
+null_if_filter_not_found = null_if_excepts(FilterNotFound)
+null_if_indexerror = null_if_excepts(IndexError)
+
+
+@null_if_indexerror
+@null_if_block_not_found
+def get_transaction_by_block_hash_and_index(eth_tester, params):
+    block_hash, transaction_index = params
+    block = eth_tester.get_block_by_hash(block_hash, full_transactions=True)
+    transaction = block['transactions'][transaction_index]
+    return transaction
+
+
+@null_if_indexerror
+@null_if_block_not_found
+def get_transaction_by_block_number_and_index(eth_tester, params):
+    block_number, transaction_index = params
+    block = eth_tester.get_block_by_number(block_number, full_transactions=True)
+    transaction = block['transactions'][transaction_index]
+    return transaction
+
+
+def create_log_filter(eth_tester, params):
+    filter_params = params[0]
+    filter_id = eth_tester.create_log_filter(**filter_params)
+    return filter_id
+
+
+def generate_random_private_key():
+    return encode_hex(bytes(bytearray((
+        random.randint(0, 255)
```

Comment:
Maybe a little note here in case anyone stumbles in, saying that this shouldn't be used for generating private keys in production, because `random` is not meant for generating secure secrets.

Reply:
Gonna add a `_` prefix and the comment.